base/range.jl

```diff
@@ -685,7 +685,7 @@ end
 
 function lerpi(j::Integer, d::Integer, a::T, b::T) where T
     @_inline_meta
-    t = j/d
+    t = eltype(T)(j)/d
```

Comment:
Why does this use `eltype` instead of just `T(j)`?

Reply:
This was required to satisfy the tests at https://github.com/JuliaLang/julia/blob/master/test/ranges.jl#L1283-L1290
In short, it ensures that things like this work:
```
julia> LinRange([0., 10.], [1., 12.], 4)
4-element LinRange{Array{Float64,1}}:
 [0.0, 10.0],[0.333333, 10.666667],[0.666667, 11.33333],[1.0, 12.0]
```

(I actually discovered such uses were possible when tests failed with an earlier version of my fix!)

Reply:
A problem is that the current implementation works even for objects that don't define `eltype(T)`: basically it just needs an algebra supporting addition and multiplication by reals.

One option would be `t = j // d` but I will wager you that there will be overflow test failures somewhere. Another option would be `t = oftype(one(T), j/d)` but amazingly even `one(::Type{Vector{T}}) where T = one(T)` isn't defined. (Why not? That seems crazy.)

Reply:
On balance perhaps the rational approach makes the most sense.

Reply:
> A problem is that the current implementation works even for objects that don't define `eltype(T)`: basically it just needs an algebra supporting addition and multiplication by reals.

That's true. Are there any examples of such types in the stdlib? If we want to support such uses, I'd like to add a specific test case for them (but I'd like to avoid defining a new type only for this purpose if possible).

Reply:
> One option would be `t = j // d` but I will wager you that there will be overflow test failures somewhere.

What kind of issues are you expecting? Overflows when `j` or `d` are huge?
It looks to me like, in your proposal, the only arithmetic operation ever performed with rational is `(1-t)`, which is merely `(d-j)//d`. I might be missing something, but I wouldn't expect any issues with that.

There might be a small performance impact though, since `lerpi` would now perform two divisions: one to compute `t`, and one to compute `1-t`. Is that a problem? Would you like me to benchmark this?

In any case, `t = j // d` seems to pass all tests (but maybe we're missing specific test cases that would trigger issues with this implementation, which is why I'm asking about anticipated issues)

Reply:
@StefanKarpinski has this annoying tendency to write tests that include the extreme ranges (joking, if he doesn't do that you can be sure that some prankster will report it as an issue).
```julia
julia> start, stop = 1, typemax(Int)
(1, 9223372036854775807)

julia> t = 3//4
3//4

julia> (1 - t) * start + t * stop
ERROR: OverflowError: 3 * 9223372036854775807 overflowed for type Int64
Stacktrace:
 [1] throw_overflowerr_binaryop(op::Symbol, x::Int64, y::Int64)
   @ Base.Checked ./checked.jl:154
 [2] checked_mul
   @ ./checked.jl:288 [inlined]
 [3] *(x::Rational{Int64}, y::Int64)
   @ Base ./rational.jl:314
 [4] top-level scope
   @ REPL[3]:1
```

Reply:
Interestingly, though, if we reversed the order of operation (divide first, multiply second) then it would be fine---though then you should check underflow for ranges like one going from `eps(0.0)` to `3*eps(0.0)`. If you can find a robust solution, it almost seems worth creating a `SmallRatio` type for this (example: https://github.com/timholy/Ratios.jl).

Reply:
Thanks, I was so focused on FP issues that hadn't thought about integer ranges.

One thing that we're able to handle in the current master is the following (which hits overflow issues with the `j//d` implementation as you mentioned):
```
julia> LinRange{Int}(1, 1+2^62, 5)
5-element LinRange{Int64}:
 1,1152921504606846976,2305843009213693952,3458764513820540928,4611686018427387904
```

(I'll note that it's currently not possible to handle `LinRange{Int}(1, typemax(Int), 3)` because `lerpi` uses FP numbers and `typemax(Int)` is not representable as a `Float64`)

<br/>

In any case, in light of this discussion, I'll try and think a bit more before proposing a new implementation that hopefully meets all envisioned use cases.

Reply:
One last issue to pay attention to is performance. I think it will be slower the more divisions you do.

Reply:
Yes, that's what I meant to say in my message above:

> There might be a small performance impact though, since `lerpi` would now perform two divisions: one to compute `t`, and one to compute `1-t`. Is that a problem? Would you like me to benchmark this?

I'll try and find an adequate compromise between all concerns involved in this.